File: languages/lua/unscramble.lua
dictionaryIterator = io.lines("/usr/share/dict/words")
input = arg[1]

if input == nil then
  os.exit()
end

function convertStringToArray(string)
  output = {}

  for index = 1, string.len(string) do
    output[index] = string.sub(string, index, index)
  end

  return output
end

function getLetterCounts(word)
  letters = convertStringToArray(word)
  counts = {}

  for index, letter in next, letters do
    letterKey = string.lower(letter)

    if counts[letterKey] ~= nil then
      counts[letterKey] = counts[letterKey] + 1
    else
      counts[letterKey] = 1
    end
  end

  return counts
end

inputLetterCounts = getLetterCounts(input)
match = nil

for word in dictionaryIterator do
  if string.len(word) == string.len(input) then
    wordLetterCounts = getLetterCounts(word)

    isMatch = true

    for letterKey, count in next, wordLetterCounts do
      if inputLetterCounts[letterKey] ~= wordLetterCounts[letterKey] then
        isMatch = false
        break
      end
    end

    if isMatch == true then
      match = word
      break
    end
  end
end

print(match)

File: languages/typescript/unscramble.ts
const fs = require("fs")
const dictionary = fs.readFileSync("/usr/share/dict/words", "utf8")
  .split("\n")

const input = process.argv[2]

if (!input) {
  process.exit(1)
}

const getLetterCounts = (word) => {
  const counts = {}

  word.split("").forEach((letter) => {
    const letterKey = letter.toLowerCase()
    if (counts[letterKey]) {
      counts[letterKey]++
    } else {
      counts[letterKey] = 1
    }
  })

  return counts
}

const inputLetterCounts = getLetterCounts(input)

const match = dictionary.find((word) => {
  if (word.length !== input.length) {
    return false
  }

  const wordLetterCounts = getLetterCounts(word)

  return Object.keys(wordLetterCounts).every((key) => (
    inputLetterCounts[key] === wordLetterCounts[key]
  ))
})

console.log(match)
